README.md
# guards #

JavaScript library for data type & data structure validations providing a
runtime analog of types. Check out [docs] for more details.

## Install ##

    npm install guards

## Require ##

    // Full version inculding String / Number / Function guards.
    var guards = require('https!raw.github.com/Gozala/guards/v0.2.0/guards.js')

    // Core version without dependencies and type specific guards.
    var guards = require('https!raw.github.com/Gozala/guards/v0.2.0/core.js')

## Example ##

    var guards = require('https!raw.github.com/Gozala/guards/v0.2.0/guards.js')
    var Point = guards.Schema({
      x: guards.Number(0),
      y: guards.Number(0)
    })

    function color(value) {
      if (typeof value === "number" && value <= 255 && value >= 0) return value
      throw new TypeError("Color is a number between 0 and 255")
    }

    var RGB = guards.Tuple([ color, color, color ])
    var Segment = guards.Schema({
      start: Point,
      end: Point,
      color: RGB,
    })

    var segment = Segment({ end: { y: 23 }, color: [17, 255, 0] })
    // { start: { x: 0, y: 0 }, end: { x: 0, y: 23 }, color: [ 17, 255, 0 ] }

## Prior art ##

- [EcmaScript Guards proposal]
- [StructsJS]

[EcmaScript Guards proposal]:http://wiki.ecmascript.org/doku.php?id=strawman:guards
[StructsJS]:https://github.com/dherman/structsjs
[docs]:http://jeditoolkit.com/guards/docs/



History.md
# Changes #

## 0.3.0 / 2011-07-02 ##

  - Removed all external dependencies.

## 0.2.0 / 2011-06-10 ##

  - Implementing in functional style to reduce boilerplate.
  - Simplified package structure.
  - Make use of graphqurie instead of bundling dependencies.
  - Splitting lib into minimal core and additional functionality.

## 0.0.2 / 2011-03-27 ##

  - Created scheme guards no longer accept array values.
  - Added `Function` guard for function validations.
  - Added `AnyOf` guard to allow simple composition of guards that accept more
    then one type / structure of values.

## 0.0.1 / 2011-03-02 ##

  - Initial release



test/test-all.js
"use strict";

var guards = require("guards");

exports["test string guard"] = function(assert) {
  var g1 = guards.String({ defaults: "default" });
  assert.equal(g1(), "default", "correct default value is assigned");
  assert.equal(g1("foo"), "foo", "string guard accepts string values");
  assert.throws(function() {
    g1(12);
  }, /String/g, "non-string passed to a string gaurd throws a TypeError");
  assert.throws(function() {
    g1(new String("foo bar"));
  }, /String/, "String instance throws exception as well");

  var g2 = guards.String();
  assert.throws(function() {
    g2();
  }, /String/, "no default value is given");

  var g3 = guards.String({ message: "Boom -> {{value}}!" });
  assert.throws(function() {
    var value = g3({});
  }, /Boom \-\> \[object Object\]!/g, "Optional error temaplate can be used");
  assert.throws(function() {
    g3();
  }, /Boom \-\>/, "no default value was given");
  assert.equal(g3("baz"), "baz", "string value validates");
};

exports["test number guard"] = function(assert) {
  var g1 = guards.Number({ defaults: 0 });
  assert.equal(g1(), 0, "correct default value is assigned");
  assert.equal(g1(17), 17, "string guard accepts string values");
  assert.throws(function() {
    g1("12");
  }, /Number/g, "non-number throws a TypeError");

  var g2 = guards.Number();
  assert.throws(function() {
    g2()
  }, "throws when no default value is given");

  var g3 = guards.Number({
    defaults: 0,
    message: "number expected not a {{type}}"
  });
  assert.throws(function() {
    g3("boom!");
  }, /number expected not a string/g, "Optional error temaplate can be used");
};

exports["test schema guard"] = function(assert) {
  var Point = guards.Schema({
    x: guards.Number({ defaults: 0 }),
    y: guards.Number({ defaults: 0 })
  });

  assert.deepEqual(Point(), { x: 0, y: 0 }, "default values are insterted");
  assert.deepEqual(Point({ x: 17, z: 50 }), { x: 17, y: 0 },
                   "non-guarded values are stripped out");
  assert.throws(function() {
    Point({ x: "5" });
  }, /Number/g, "Wrong type in caught by nested guard");

  assert.throws(function() {
    Point("{ y: 6 }");
  }, /Object/g, "Scheme guard throws on non-object values");


  var Segment = guards.Schema({
    start: Point,
    end: Point,
    opacity: guards.Number({ defaults: 1 })
  });

  assert.deepEqual(Segment(),
                  { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                  "default values are used by all nested guards");

  assert.deepEqual(Segment({ end: { x: 17, foo: 'bar' }, opacity: 0.5 }),
                  { start: { x: 0, y: 0 }, end: { x: 17, y: 0 }, opacity: 0.5 },
                  "fallback to default values and strip out non-guarded ones");

  assert.throws(function() {
    Segment({ start: 17 })
  }, /Object/g, "nested guard throw on wrong value types");
};

exports["test array guard"] = function(assert) {
  var Words = guards.Array(guards.String({ defaults: "" }));

  assert.deepEqual(Words([]), [], "Empty array passes through");
  assert.deepEqual(Words([ "foo", "bar" ]), [ "foo", "bar" ],
                   "primitive type array works fine");

  assert.throws(function() {
    Words([ "foo", 9 ]);
  }, /String/g, "throw because it got number instead of string");

  var Point = guards.Schema({
    x: guards.Number({ defaults: 0 }),
    y: guards.Number({ defaults: 0 })
  });
  var Points = guards.Array(Point);

  assert.deepEqual(Points([{ foo: 'bar' }, { x: 2, y: 8 }]),
                   [ { x: 0, y: 0 }, { x: 2, y: 8 } ],
                   "Defualts used were needed and non-guradede stripped off");

  assert.throws(function() {
    Points({ x: 2, y: 8 });
  }, /Array/g, "TypeError is thrown if value is not an array");


  var Graph = guards.Array(Points);

  assert.deepEqual(Graph([]), [], "empty array passes through");
  assert.deepEqual(Graph([
                          [{ x: 17, foo: "bar" }, { x: 16 }],
                          [{ y: 4 }],
                          []
                   ]),
                   [
                     [ { x: 17, y: 0 }, { x: 16, y: 0 } ],
                     [ { x: 0, y: 4 } ],
                     []
                   ],
                   "arrray of arrays of points works fine");
};

exports["test tuple guards"] = function(assert) {
  var guards = require("guards");
  var Point = guards.Schema({
    x: guards.Number({ defaults: 0 }),
    y: guards.Number({ defaults: 0 })
  });
  var Segment = guards.Schema({
    start: Point,
    end: Point,
    opacity: guards.Number({ defaults: 1 })
  });
  var Triangle = guards.Tuple([ Segment, Segment, Segment ]);

  assert.deepEqual(Triangle(),
                  [
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                  ],
                  "if no argument passed then defaults are picked up");

  assert.deepEqual(Triangle([
                    { opacity: 0, foo: "bar" },
                    { start: { x: 2 } }
                  ]),
                  [
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 0 },
                    { start: { x: 2, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                  ],
                  "defualts picked up non-guarded stripped");

    assert.throws(function() {
      Triangle("foo");
    }, /Array/g, "Tuple guard accepts only array");

    assert.throws(function() {
      Triangle([{ start: { x: "3" } } ]);
    }, /Number/g, "Nested guard throws if wrong `value` is passed");


    var Pointer = guards.Tuple([ Point, Segment ]);

    assert.deepEqual(Pointer(),
                    [
                      { x: 0, y: 0 },
                      { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                    ],
                    "Falls back to the default values");

    assert.deepEqual(Pointer([{ x: 17 }, { opacity: 0 }]),
                    [
                      { x: 17, y: 0 },
                      { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 0 }
                    ],
                    "Defaults picked up, non-guarded stripped out");


    assert.deepEqual(Pointer([{ foo: "bar" }, { baz: "bla" }, "foo"]),
                    [
                      { x: 0, y: 0 },
                      { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                    ],
                    "Non-guarded elements and values ar stripped out");
};

exports["test custom guards"] = function(assert) {
  var Point = guards.Schema({
    x: guards.Number({ defaults: 0 }),
    y: guards.Number({ defaults: 0 })
  });
  function color(value) {
    if (typeof value === "number" && value <= 255 && value >= 0)
      return value
    throw new TypeError("Color is a number between 0 and 255");
  }
  var RGB = guards.Tuple([ color, color, color ]);
  var Segment = guards.Schema({
    start: Point,
    end: Point,
    color: RGB,
  });

  assert.deepEqual(Segment({ end: { y: 23 }, color: [17, 255, 0] }),
                  { start: { x: 0, y: 0 },
                    end: { x: 0, y: 23 },
                    color: [ 17, 255, 0 ]
                  },
                  "defaults values used for missing properties");
  assert.deepEqual(Segment({ color: [ 1, 2, 3, 0.5 ] }).color,
                  [ 1, 2, 3 ],
                  "non guarded tuple elements are stripped out");

  assert.throws(function() {
    Segment({ colors: [] });
  }, /Color is a number between 0 and 255/, "no defualt for colors");
  assert.throws(function() {
    Segment({ colors: [ 0, 255, 256 ] });
  }, /Color is a/, "color out of the range");
};

exports["test AnyOf guards"] = function(assert) {
  var guards = require("guards");
  var ObjectPoint = guards.Schema({
    x: guards.Number({ defaults: 0 }),
    y: guards.Number({ defaults: 0 })
  });
  var ArrayPoint = guards.Tuple([
    guards.Number({ defaults: 0 }),
    guards.Number({ defaults: 0 })
  ]);
  var Point = guards.AnyOf(ObjectPoint, ArrayPoint);

  assert.deepEqual(Point([ 1 ]), [ 1, 0 ], "ArrayPoint validates");
  assert.deepEqual(Point({ y: 15 }), { x: 0, y: 15 }, "ObjectPoint validates");
  assert.throws(function() {
    Point(1, 2)
  }, /invalid/, "Number does not validates");
};

exports["test function guards"] = function(assert) {
  var guards = require("guards");
  var Callee = guards.Function({ defaults: "Anonymous" });
  var f1 = function () { return "hello world" }

  assert.equal(Callee(Object), Object, "Object validates");
  assert.equal(Callee(f1), f1, "function validates");
  assert.throws(function() {
    Callee(7);
  }, /Function expected/, "Number does not validates as Function");
  assert.throws(function() {
    var Callee2 = guards.Function({
      defaults: "Hi",
      message: "{{type}} is not a function."
    });
    Callee2({});
  }, /object is not a function/, "custom error message is thrown");
};

require("test").run(exports)


tests/test-guards.js
/* vim:set ts=2 sw=2 sts=2 expandtab */
/*jshint undef: true es5: true node: true devel: true
         forin: true */
/*global define: true */

(typeof define !== "function" ? function($){ $(require, exports, module); } : define)(function(require, exports, module, undefined) {

"use strict";

var guards = require("../guards.js");

exports["test string guard"] = function(assert) {
  var g1 = guards.String("default");
  assert.equal(g1(), "default", "correct default value is assigned");
  assert.equal(g1("foo"), "foo", "string guard accepts string values");
  assert.throws(function() {
    g1(12);
  }, /String/g, "non-string passed to a string gaurd throws a TypeError");
  assert.throws(function() {
    g1(new String("foo bar"));
  }, /String/, "String instance throws exception as well");

  var g2 = guards.String();
  assert.throws(function() {
    g2();
  }, /String/, "no default value is given");

  var g3 = guards.String(undefined, "Boom -> {{value}}!");
  assert.throws(function() {
    var value = g3({});
  }, /Boom \-\> \[object Object\]!/g, "Optional error temaplate can be used");
  assert.throws(function() {
    g3();
  }, /Boom \-\>/, "no default value was given");
  assert.equal(g3("baz"), "baz", "string value validates");
};

exports["test number guard"] = function(assert) {
  var g1 = guards.Number();
  assert.equal(g1(12), 12, "numbers work just fine");
  assert.throws(function() {
    g1('12');
  }, /Number/, "non-number throws");
  assert.throws(function() {
    g1();
  }, /Number/, "undefined throws without defaults attribute");
  var g2 = guards.Number(0);
  assert.equal(g2(), 0, "correct default value is assigned");
  assert.equal(g2(17), 17, "string guard accepts string values");
  assert.throws(function() {
    g2("12");
  }, /Number/g, "non-number throws a TypeError");

  var g3 = guards.Number(0, "number expected not a {{type}}");
  assert.throws(function() {
    g3("boom!");
  }, /number expected not a string/g, "Optional error temaplate can be used");
};

exports["test schema guard"] = function(assert) {
  var Point = guards.Schema({
    x: guards.Number(0),
    y: guards.Number(0)
  });

  assert.deepEqual(Point(), { x: 0, y: 0 }, "default values are insterted");
  assert.deepEqual(Point({ x: 17, z: 50 }), { x: 17, y: 0 },
                   "non-guarded values are stripped out");
  assert.throws(function() {
    Point({ x: "5" });
  }, /Number/g, "Wrong type in caught by nested guard");

  assert.throws(function() {
    Point("{ y: 6 }");
  }, /Object/g, "Scheme guard throws on non-object values");


  var Segment = guards.Schema({
    start: Point,
    end: Point,
    opacity: guards.Number(1)
  });

  assert.deepEqual(Segment(),
                  { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                  "default values are used by all nested guards");

  assert.deepEqual(Segment({ end: { x: 17, foo: 'bar' }, opacity: 0.5 }),
                  { start: { x: 0, y: 0 }, end: { x: 17, y: 0 }, opacity: 0.5 },
                  "fallback to default values and strip out non-guarded ones");

  assert.throws(function() {
    Segment({ start: 17 })
  }, /Object/g, "nested guard throw on wrong value types");
};

exports["test array guard"] = function(assert) {
  var array = guards.Array();
  assert.deepEqual(array([ 1, 2, 3 ]), [ 1, 2, 3 ], "array returned");
  assert.throws(function() {
    array(1, 2, 3);
  }, /Array/, "throw because non-array passed");

  var arraymsg = guards.Array(undefined, "Boom!");
  assert.deepEqual(array([ 1, 2, 3 ]), [ 1, 2, 3 ], "array returned");
  assert.throws(function() {
    arraymsg(1, 2, 3);
  }, /Boom/, "throw because non-array passed");

  var Words = guards.Array(guards.String(''));

  assert.deepEqual(Words([]), [], "Empty array passes through");
  assert.deepEqual(Words([ "foo", "bar" ]), [ "foo", "bar" ],
                   "primitive type array works fine");

  assert.throws(function() {
    Words([ "foo", 9 ]);
  }, /String/g, "throw because it got number instead of string");

  var Point = guards.Schema({
    x: guards.Number(0),
    y: guards.Number(0)
  });
  var Points = guards.Array(Point);

  assert.deepEqual(Points([{ foo: 'bar' }, { x: 2, y: 8 }]),
                   [ { x: 0, y: 0 }, { x: 2, y: 8 } ],
                   "Defualts used were needed and non-guradede stripped off");

  assert.throws(function() {
    Points({ x: 2, y: 8 });
  }, /Array/g, "TypeError is thrown if value is not an array");


  var Graph = guards.Array(Points);

  assert.deepEqual(Graph([]), [], "empty array passes through");
  assert.deepEqual(Graph([
                          [{ x: 17, foo: "bar" }, { x: 16 }],
                          [{ y: 4 }],
                          []
                   ]),
                   [
                     [ { x: 17, y: 0 }, { x: 16, y: 0 } ],
                     [ { x: 0, y: 4 } ],
                     []
                   ],
                   "arrray of arrays of points works fine");
};

exports["test tuple guards"] = function(assert) {
  var guards = require("../guards.js");
  var Point = guards.Schema({
    x: guards.Number(0),
    y: guards.Number(0)
  });
  var Segment = guards.Schema({
    start: Point,
    end: Point,
    opacity: guards.Number(1)
  });
  var Triangle = guards.Tuple([ Segment, Segment, Segment ]);

  assert.deepEqual(Triangle(),
                  [
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                  ],
                  "if no argument passed then defaults are picked up");

  assert.deepEqual(Triangle([
                    { opacity: 0, foo: "bar" },
                    { start: { x: 2 } }
                  ]),
                  [
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 0 },
                    { start: { x: 2, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
                    { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                  ],
                  "defualts picked up non-guarded stripped");

    assert.throws(function() {
      Triangle("foo");
    }, /Array/g, "Tuple guard accepts only array");

    assert.throws(function() {
      Triangle([{ start: { x: "3" } } ]);
    }, /Number/g, "Nested guard throws if wrong `value` is passed");


    var Pointer = guards.Tuple([ Point, Segment ]);

    assert.deepEqual(Pointer(),
                    [
                      { x: 0, y: 0 },
                      { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                    ],
                    "Falls back to the default values");

    assert.deepEqual(Pointer([{ x: 17 }, { opacity: 0 }]),
                    [
                      { x: 17, y: 0 },
                      { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 0 }
                    ],
                    "Defaults picked up, non-guarded stripped out");


    assert.deepEqual(Pointer([{ foo: "bar" }, { baz: "bla" }, "foo"]),
                    [
                      { x: 0, y: 0 },
                      { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
                    ],
                    "Non-guarded elements and values ar stripped out");
};

exports["test custom guards"] = function(assert) {
  var Point = guards.Schema({
    x: guards.Number(0),
    y: guards.Number(0)
  });
  function color(value) {
    if (typeof value === "number" && value <= 255 && value >= 0)
      return value
    throw new TypeError("Color is a number between 0 and 255");
  }
  var RGB = guards.Tuple([ color, color, color ]);
  var Segment = guards.Schema({
    start: Point,
    end: Point,
    color: RGB,
  });

  assert.deepEqual(Segment({ end: { y: 23 }, color: [17, 255, 0] }),
                  { start: { x: 0, y: 0 },
                    end: { x: 0, y: 23 },
                    color: [ 17, 255, 0 ]
                  },
                  "defaults values used for missing properties");
  assert.deepEqual(Segment({ color: [ 1, 2, 3, 0.5 ] }).color,
                  [ 1, 2, 3 ],
                  "non guarded tuple elements are stripped out");

  assert.throws(function() {
    Segment({ colors: [] });
  }, /Color is a number between 0 and 255/, "no defualt for colors");
  assert.throws(function() {
    Segment({ colors: [ 0, 255, 256 ] });
  }, /Color is a/, "color out of the range");
};

exports["test AnyOf guards"] = function(assert) {
  var guards = require("../guards.js");
  var ObjectPoint = guards.Schema({
    x: guards.Number(0),
    y: guards.Number(0)
  });
  var ArrayPoint = guards.Tuple([ guards.Number(0), guards.Number(0) ]);
  var Point = guards.AnyOf([ ObjectPoint, ArrayPoint ]);

  assert.deepEqual(Point([ 1 ]), [ 1, 0 ], "ArrayPoint validates");
  assert.deepEqual(Point({ y: 15 }), { x: 0, y: 15 }, "ObjectPoint validates");
  assert.throws(function() {
    Point(1, 2)
  }, /invalid/, "Number does not validates");
};

exports["test function guards"] = function(assert) {
  var guards = require("../guards.js");
  var Callee = guards.Function();
  var f1 = function () { return "hello world" }

  assert.equal(Callee(Object), Object, "Object validates");
  assert.equal(Callee(f1), f1, "function validates");
  assert.throws(function() {
    Callee(7);
  }, /Function expected/, "Number does not validates as Function");
  assert.throws(function() {
    var Callee2 = guards.Function(undefined, "{{type}} is not a function.");
    Callee2({});
  }, /object is not a function/, "custom error message is thrown");
};

require("test").run(exports)

});


lib/extendables.js
/* vim:set ts=2 sw=2 sts=2 expandtab */
/*jshint undef: true es5: true node: true devel: true
         forin: true latedef: false supernew: true */
/*global define: true */

(typeof define !== "function" ? function($){ $(require, exports, module); } : define)(function(require, exports, module, undefined) {

"use strict";

function getOwnPropertyDescriptors(object) {
  var descriptors = {};
  Object.getOwnPropertyNames(object).forEach(function(name) {
    descriptors[name] = Object.getOwnPropertyDescriptor(object, name);
  });
  return descriptors;
}

function supplement(target, source) {
  var descriptors = {};
  var names = Object.getOwnPropertyNames(target);
  Object.getOwnPropertyNames(source).forEach(function(name) {
    if (!~names.indexOf(name)) {
      descriptors[name] = Object.getOwnPropertyDescriptor(source, name);
    }
  });
  return Object.defineProperties(target, descriptors);
}

function Constructor(base) {
  return function Extendable() {
    var value, extendable = this;
    if (!(extendable instanceof Extendable))
        extendable = Object.create(Extendable.prototype);

    value = base.apply(extendable, arguments);
    return value === undefined ? extendable : value;
  }
}
function Extendable() {
  return this instanceof Extendable ? this : Object.create(Extendable.prototype);
}

Object.defineProperties(Extendable, {
  extend: {
    value: function extend(source) {
      var constructor, descriptors = getOwnPropertyDescriptors(source || {});
      // If `constructor` is not defined by `source` then we generate a default
      // `constructor` that delegates to the `constructor` of the base class.
      if (typeof descriptors.constructor !== "object")
        descriptors.constructor = { value: new Constructor(this) };
      // Copying all the non-existing properties to the decedent.
      constructor = supplement(descriptors.constructor.value, this);
      // Override prototype of the decedent.
      constructor.prototype = Object.create(this.prototype, descriptors);
      return constructor;
    },
    enumerable: true
  }
});

exports.Extendable = Extendable;
exports.version = "0.1.2";

});


core.js
/* vim:set ts=2 sw=2 sts=2 expandtab */
/*jshint asi: true undef: true es5: true node: true devel: true
         forin: false */
/*global define: true */

(typeof define !== "function" ? function($){ $(require, exports, module); } : define)(function(require, exports, module, undefined) {

var slice = Array.prototype.slice
var stringify = Object.prototype.toString

function compose() {
  var funcs = slice.call(arguments)
  return function composed() {
    var args = slice.call(arguments)
    var i = funcs.length
    while (0 <= --i) args = [ funcs[i].apply(this, args) ]
    return args[0]
  }
}
var isArray = Array.isArray || function isArray(value) {
  return stringify.call(value) === '[object Array]'
}
function isSchema(value) {
  return value && typeof value === 'object' && !isArray(value)
}
function reference(value) { return value }

exports.version = "0.2.0";

/**
 * # Guard #
 *
 * Function takes `isValid` function `defaults` function or value and `message`
 * as an argument and returns guard function that may be used to guard values.
 * Each value passed to the guard will be validated using given `isValid` if it
 * returns `true` value is passed through, otherwise `TypeError` is thrown with
 * a given message. If guard is called without an argument, `defaults` is used.
 * If defaults is a function it's value is used otherwise `defaults` is used
 * itself.
 * @param {Function} isValid
 *    Function that will be used by guard to validate values. It will be called
 *    with `value` argument that needs to be validated. If function returns
 *    `true` value is valid if `false` it's not.
 * @param {Function|*} defaults
 *    Function that returns default value or a default value itself to be used
 *    when guard is called without arguments.
 * @param {String} [message="Unexpected value: `{{value}}`"]
 *    Optional `message` argument may be passed that will be used as a template
 *    for a `TypeError` message that is thrown by generated guards when called
 *    with invalid values.
 */

/**
 * # Guard #
 *
 * Guard function that may be called with a `value` to be set to a guarded
 * variable. If `value` is invalid `TypeError` is thrown. Optionally second
 * `name` argument may be passed, which is useful when guards are used for
 * object properties. In such case `name` argument just name of object
 * property and will be used to give a better error messages.
 * @param {Object|String|Number|function} value
 *    Value to be validated.
 * @param {String} name
 *    Name of the property that is being guarded.
 */
function Guard(isValid, defaults, message) {
  message = message || "Unexpected value: `{{value}}`"
  return function guard(value, name) {
    value = value !== undefined ? value :
            typeof defaults === 'function' ? defaults() : defaults
    if (!isValid(value))
       throw new TypeError(message.replace("{{name}}", name).
                                   replace("{{value}}", value).
                                   replace("{{type}}", typeof value));
    return value
  }
}
exports.Guard = Guard

/**
 * # Schema #
 *
 * Schema is useful for defining guards for data objects that have particular
 * structure. Function takes `descriptor` argument that is a map of guards
 * guarding same named properties of the value being validated. Scheme may
 * contain guards for a primitive values like `String` and `Number` and also
 * guards for more complex data structures defined by other `Schema`s, or to
 * put it other way `Schema` may contain guards that were created by `Schema`
 * itself which allows defining deeply nested data structures.
 *
 * Generated guard will accept only object `values` as an argument. All the
 * non-guarded properties (that are not present in the `descriptor`) of the
 * `value` will be stripped out. All the missing properties of the `value`
 * will be replaced / assembled from the defaults if associated guards provide
 * fallback mechanism to default value.
 *
 * @param {Object} descriptor
 *    Object containing guards for the associated (same named properties) of
 *    the guarded object `value`.
 * @param {String} [message]
 *    Optional error message template that will be a message of a `TypeError`
 *    that is will be thrown if returned guard is invoked with a wrong `value`
 *    type (other then "object" or "undefined"). If `message` contains
 *    `"{{value}}"` and `"{{type}}"` strings they are going to be replaced with
 *    an actual `value` and it's type.
 *
 * ## Examples ##
 *
 *      var guards = require("guards");
 *      var Point = guards.Schema({
 *        x: guards.Number(0),
 *        y: guards.Number(0)
 *      });
 *
 *      new Point
 *      // { x: 0, y: 0 }
 *
 *      Point({ x: 17, z: 50 })
 *      // { x: 17, y: 0 }
 *
 *      Point({ x: "5" })
 *      // TypeError: Number expected instead of string `5`
 *
 *      Point("{ y: 6 }")
 *      // TypeError: Object expected instead of string `{ y: 6 }`
 *
 *
 *      var Segment = guards.Schema({
 *        start: Point,
 *        end: Point,
 *        opacity: guards.Number(1)
 *      })
 *
 *      new Segment
 *      // { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
 *
 *      Segment({ end: { x: 17 }, opacity: 0.5 })
 *      // { start: { x: 0, y: 0 }, end: { x: 17, y: 0 }, opacity: 0.5 }
 *
 *      Segment({ start: 17 });
 *      // TypeError: Object expected instead of number `17`
 *
 */
exports.Schema = function Schema(schema, message) {
  message = message || 'Object expected instead of {{type}} `{{value}}`'
  return compose(function guard(value) {
    var validated = {}
    for (var key in schema) validated[key] = schema[key](value[key], key)
    return validated
  }, Guard(isSchema, Object, message))
}

/**
 * # Array #
 *
 * Array can be used to define guards for an arrays containing elements of some
 * type or schema. Function takes `guard` as an argument that will guard all the
 * elements of the `value` array that is passed to the returned guard.
 *
 * @param {Function} guard
 *    Guard that is going to be used to verify elements of the array. It can
 *    be any guard created by `String`, `Schema`, `Array` or any custom guard
 *    as well.
 * @param {String} [message]
 *    Optional error message template that will be a message of a `TypeError`
 *    that is will be thrown if returned guard is invoked with a wrong `value`
 *    type (other then "array" or "undefined"). If `message` contains
 *    `"{{value}}"` and `"{{type}}"` strings they are going to be replaced with
 *    an actual `value` and it's type.
 *
 * ## Examples ##
 *
 *      var guards = require("guards")
 *      var Words = guards.Array(guards.String(''))
 *
 *      Words([ "foo", "bar" ])
 *      // [ 'foo', 'bar' ]
 *
 *      Words([ "foo", 9 ])
 *      // TypeError: String expected instead of number `9`
 *
 *      var Point = guards.Schema({
 *        x: guards.Number(0),
 *        y: guards.Number(0)
 *      })
 *      var Points = guards.Array(Point)
 *
 *      new Points([{}, { x: 2, y: 8 }])
 *      // [ { x: 0, y: 0 }, { x: 2, y: 8 } ]
 *
 *      Points({ x: 2, y: 8 })
 *      // TypeError: Array expected instead of object `[object Object]`
 *
 *
 *      var Graph = guards.Array(Points)
 *
 *      new Graph
 *      // []
 *
 *      Graph([
 *        [{ x: 17, foo: "bar" }, { x: 16 }],
 *        [{ y: 4 }],
 *        []
 *      ])
 *      // [ [ { x: 17, y: 0 }, { x: 16, y: 0 } ], [ { x: 0, y: 4 } ], [] ]
 */
exports.Array = function Array(type, message) {
  var Type = [].constructor
  message = message || 'Array expected instead of {{type}} `{{value}}`'
  type = type || reference
  return compose(function guard(value) {
    return value.map(type)
  }, Guard(isArray, Type, message))
}


/**
 * # Tuple #
 *
 * Tuple can be used to define guards for an arrays containing predefined
 * amount of elements guarded by specific guards. Tuple guards are something
 * in between Array and Schema guards. Function takes array of guards as an
 * argument that will be used to validate same indexed elements of the `value`
 * array that is passed to the returned guard.
 *
 * @param {Function[]} guards
 *    Guards that are going to be used to verify elements of the array. It can
 *    be any guard created by `String`, `Schema`, `Array` or any custom guard
 *    as well.
 * @param {String} [message]
 *    Optional error message template that will be a message of a `TypeError`
 *    that is will be thrown if returned guard is invoked with a wrong `value`
 *    type (other then "array" or "undefined"). If `message` contains
 *    `"{{value}}"` and `"{{type}}"` strings they are going to be replaced with
 *    an actual `value` and it's type.
 *
 * ## Examples ##
 *
 *      var guards = require("guards");
 *      var Point = guards.Schema({
 *        x: guards.Number(0),
 *        y: guards.Number(0)
 *      });
 *      var Segment = guards.Schema({
 *        start: Point,
 *        end: Point,
 *        opacity: guards.Number(1)
 *      });
 *      var Triangle = guards.Tuple([ Segment, Segment, Segment ]);
 *
 *      var t1 = Triangle();
 *      // [ { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
 *      //   { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
 *      //   { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
 *      // ]
 *
 *      var t2 = Triangle([
 *        { opacity: 0, foo: "bar" },
 *        { start: { x: 2 } }
 *      ]);
 *      // [ { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 0 },
 *      //   { start: { x: 2, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 },
 *      //   { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 }
 *      // ]
 *
 *      var t2 = Triangle("foo");
 *      // TypeError: Array expected instead of string `foo`
 *
 *      var t3 = Triangle([{ start: { x: '3' } } ]);
 *      // TypeError: Number expected instead of string `3`
 *
 *      var Pointer = guards.Tuple([ Point, Segment ]);
 *
 *      var p1 = Pointer();
 *      // [ { x: 0, y: 0 }, { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 } ]
 *
 *      var p2 = Pointer([ { x: 17 }, { opacity: 0 } ]);
 *      // [ { x: 17, y: 0 }, { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 0 } ]
 *
 *      var p3 = Pointer([ { foo: "bar" }, { baz: "bla" }, "foo" ]);
 *      // [ { x: 0, y: 0 }, { start: { x: 0, y: 0 }, end: { x: 0, y: 0 }, opacity: 1 } ]
 */
exports.Tuple = function Tuple(schema, message) {
  message = message || 'Array expected instead of {{type}} `{{value}}`'
  return compose(function guard(value) {
    return schema.map(function(guard, index) {
      return guard(value[index], index)
    })
  }, Guard(isArray, Array, message))
}

/**
 * # AnyOf #
 *
 * `AnyOf` can be used to define guards that validates `value`s that must
 * satisfy just one of many guards. This is handy in specific specific
 * scenarios were valid `value` may have different types or schemas.
 * Function takes any number guards as an arguments and returns composed
 * guard, which when called will try to validate a given `value` with a given
 * guards in an order they were passed, the first validate `value` is returned
 * as result, unless non will validate in which case `TypeError` is thrown.
 *
 * @params {Function} guard
 *    Guards used for validations.
 * @returns {Function}
 *
 * ## Examples ##
 *
 *      var guards = require("guards")
 *      var ObjectPoint = guards.Schema({
 *        x: guards.Number(0),
 *        y: guards.Number(0)
 *      })
 *      var ArrayPoint = guards.Tuple.extend([
 *        guards.Number(0),
 *        guards.Number(0)
 *      ])
 *      var Point = guards.AnyOf([ ObjectPoint, ArrayPoint ])
 *
 *      Point([ 1 ])
 *      // [ 1, 0 ]
 *
 *      Point({ y: 15 })
 *      // { x: 0, y: 15 }
 *
 *      Point(1, 2)
 *      // TypeError: Passed value: `1` has invalid type or structure
 */
exports.AnyOf = function AnyOf(guards, message) {
  message = message || 'Passed value: `{{value}}` has invalid type or structure'
  return function guard(value, name) {
    var i = 0, ii = guards.length
    while(i < ii) {
      try { return guards[i++](value, name) } catch (error) { /* swallow */ }
    }
    throw new TypeError(message.replace("{{name}}", name).
                                replace("{{value}}", value).
                                replace("{{type}}", typeof value))
  }
}

});

